presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java

```diff
@@ -328,8 +327,6 @@ PlanNodeStatsEstimate addJoinComplementStats(
             PlanNodeStatsEstimate baseJoinStats,
             PlanNodeStatsEstimate joinComplementStats)
     {
-        checkState(baseJoinStats.getSymbolsWithKnownStatistics().containsAll(joinComplementStats.getSymbolsWithKnownStatistics()));
```

Comment:
Please try changing it to:
```
checkState(sourceJoinStats.getSymbolsWithKnownStatistics().containsAll(joinComplementStats.getSymbolsWithKnownStatistics()));
```
Join complement stats shouldn't have stats for more symbols (can they?) than stats from which it was computed, no?

Reply:
Now it is very obscure. `PlanNodeStatsEstimate#getSymbolStatistics` returns `UNKNOWN` if stats are missing. `normalize` removes `UNKNOWN`. Until we refactor this, and make it more clear, it is safer to do not have this check.

Reply:
Take a look at:
https://github.com/prestodb/presto/pull/11516/commits/55e906ad2ea4c13d238070be14de59353d091d6a#diff-09acb46da127b30c7dc4632cf5052c1aR279
It's clear that complement stats should be derived from base stats (outer stats). Therefore:
```
checkState(sourceJoinStats.getSymbolsWithKnownStatistics().containsAll(joinComplementStats.getSymbolsWithKnownStatistics()));
```
should be fine. You can remove it, but it makes reading code easier later on as it simplifies reasoning.

> Now it is very obscure. PlanNodeStatsEstimate#getSymbolStatistics returns UNKNOWN if stats are missing. normalize removes UNKNOWN. Until we refactor this, and make it more clear, it is safer to do not have this check.

That's a design flaw. stats should keep stats (even unknown) for all symbols that are produced by plan node. We should assert that plan node output symbols match stats symbols then.

Reply:
I would really love to do not introduce any additional checks. Although it seems that now this check will always pass, i can't be sure that some allegedly "unrelated" change will break it again.